src/main/java/org/codehaus/mojo/flatten/cifriendly/CiInterpolatorImpl.java

```diff
@@ -0,0 +1,351 @@
+
+	                            if ( value != null && value.toString().contains( wholeExpr ) )
+	                            {
+	                                bestAnswer = value;
+	                                value = null;
+	                            }
+	                        }
+
+	                        // this is the simplest recursion check to catch exact recursion
+	                        // (non synonym), and avoid the extra effort of more string
+	                        // searching.
+	                        if ( value == null && bestAnswer != null )
+	                        {
+	                            throw new InterpolationCycleException( recursionInterceptor, realExpr, wholeExpr );
+	                        }
+
+	                        if ( value != null )
+	                        {
+	                            value = interpolate( String.valueOf( value ), recursionInterceptor, unresolvable );
+
+	                            if ( postProcessors != null && !postProcessors.isEmpty() )
+	                            {
+	                                for ( InterpolationPostProcessor postProcessor : postProcessors )
+	                                {
+	                                    Object newVal = postProcessor.execute( realExpr, value );
+	                                    if ( newVal != null )
+	                                    {
+	                                        value = newVal;
+	                                        break;
+	                                    }
+	                                }
+	                            }
+
+	                            // could use:
+	                            // result = matcher.replaceFirst( stringValue );
+	                            // but this could result in multiple lookups of stringValue, and replaceAll is not correct
+	                            // behaviour
+	                            result.append( String.valueOf( value ) );
+	                            resolved = true;
+	                        }
+	                        else
+	                        {
+	                            unresolvable.add( wholeExpr );
+	                        }
+	                    }
+	                    finally
+	                    {
+	                        recursionInterceptor.expressionResolutionFinished( realExpr );
+	                    }
+	                }
+
+	                if ( !resolved )
+	                {
+	                    result.append( wholeExpr );
+	                }
+
+	                if ( endIdx > -1 )
+	                {
+	                    endIdx += endExpr.length() - 1;
+	                }
+	            }
+
+	            if ( endIdx == -1 && startIdx > -1 )
+	            {
+	                result.append( input, startIdx, input.length());
+	            }
+	            else if ( endIdx < input.length() )
+	            {
+	                result.append( input, endIdx + 1, input.length() );
+	            }
+
```

Comment:
woho. Can't this ~120 lines of code be reduced to a view lines using regex, findFirst and replace? In the end you are only looking for 3 fixded given expressions. Also you could directly indexOf or replace for that 3 variables if they are hardcoded anyhow... Just looks overcomplicated to me. I might also be missing something. Just a suggestion, nothing that could not be accepted as is. A regular JUnit for this would be helpful and more explanatory as well...

Reply:
CiInterpolatorImpl just basically replicates StringSearchInterpolator from plexus-interpolation with small changes so that it only replaces those three variables. It is ugly, but works with the same behavior as the resolving of other variables in the plugin. Using regex might detect and resolve circular and nested variables differently.

Reply:
Perfectly fine then. Get the point. Maybe a link to the "copied" implementation (StringSearchInterpolator) in the javadoc (`{@link package.name.StringSearchInterpolator}`) with such hint would be lovely.